src/balancer/BalancerController.sol

```diff
@@ -177,21 +177,34 @@ contract BalancerController is IController {
         returns (bool, address[] memory, address[] memory)
     {
         (
-            ,
+            IVault.SwapKind kind,
             IVault.BatchSwapStep[] memory swaps,
             IAsset[] memory assets,
             ,
             ,
         ) = abi.decode(data, (
-                uint8, IVault.BatchSwapStep[], IAsset[], IVault.FundManagement, uint256[], uint256
+                IVault.SwapKind,
+                IVault.BatchSwapStep[],
+                IAsset[],
+                IVault.FundManagement,
+                uint256[],
+                uint256
             )
         );
 
         if (!isMultiHopSwap(swaps))
             return (false, new address[](0), new address[](0));
 
-        uint tokenInIndex = swaps[swaps.length - 1].assetOutIndex;
-        uint tokenOutIndex = swaps[0].assetInIndex;
+        uint tokenInIndex;
+        uint tokenOutIndex;
+
+        if (kind == IVault.SwapKind.GIVEN_IN) {
+            tokenInIndex = swaps[swaps.length - 1].assetOutIndex;
+            tokenOutIndex = swaps[0].assetInIndex;
+        } else {
+            tokenOutIndex = swaps[swaps.length - 1].assetInIndex;
+            tokenInIndex = swaps[0].assetOutIndex;
+        }
```

Comment:
I think this is backwards. If I understand correctly, it should be:

GIVEN IN
- in = swaps[0].assetIn
- out = swaps[last].assetOut

GIVEN OUT
- in = swaps[last].assetIn
- out = swaps[0].assetOut

In other words, assetIn and assetOut still mean in and out, but you just order the swaps in the other order so that there's always an amount on the first swap.

I'd recommend you write a forking test to make sure it passes though, as the documentation isn't totally clear.

Reply:
balancer uses assetIn and assetOut differently, assetIn -> asset coming into the vault, assetOut -> asset coming out of the vault.

Reply:
Yep, you're 100% right. Thanks for clarifying.